### PR TITLE
[Rebalancer] support partition rebalancer

### LINKED ISSUE #4845: [Proposal] suppoort PartitionRebalancer by using TwoDimensionalGreedyAlgo
**Is your feature request related to a problem? Please describe.**
After #4763, we want to import the kudu's rebalance strategy [TwoDimensionalGreedyAlgo](https://github.com/apache/kudu/blob/c7bf1c140e00ba8895ec6a14fa918c190dff2e1f/src/kudu/rebalance/rebalance_algo.h#L139-L147).
The kudu's two dims are `cluster` & `table`. Two dims of Doris should be `cluster` & `partition`.
And we only consider about the replica count, do not consider replica size.

**Describe the solution you'd like(POC)**
```
PartitionRebalancer:
    @override
    selectAlternativeTabletsForCluster:
        // Cleanup expired moves, or we can move the cleanup into updateLoadStatistic()
        movesInProgress.cleanUp()

        TwoDimensionalGreedyAlgo.CheckMovesCompleted(movesInProgress)

        // GetNextMoves gens vector of move{partitionId, indexId, fromBE, destBE}
        moves = TwoDimensionalGreedyAlgo.getNextMoves(clusterBalanceInfo, movesInProgress, maxMoveNum)

        // The TabletSchedCtx which will add to TabletScheduler must have to specify the tablet id
        for move in moves:
            // Pick a tablet of [partitionId-indexId] on fromBE randomly 
            tabletId = randomPickFrom(move)
            // srcReplica & destBe will be specified later, in completeSchedCtx()
            tabletSchedCtx = generateSchedCtx(tabletId)
            alternativeTablets.add(tabletSchedCtx)
            // Mark this move as in progress, the map is <tabletId, {Move, ToDeleteReplicaId}>, the ToDeleteReplicaId is only for func getToDeleteReplicaId()
            movesInProgress.add(tabletId, {move, -1})

    @override
    completeSchedCtx:
        try:
            move = movesInProgress.get(tabletId).move

            // ClusterBalanceInfo may be updated, the move may not improve the partition's or cluster's balance. So we should check the move's validation.
            checkValidation(move, clusterBalanceInfo)

            // Specify srcReplica & destBe, consider about PathSlots ...
            // Omitted ...

            movesInProgress.put(tabletId, {move, srcReplicaId})
        except e:
            // Can't create balance task for this tablet
            movesInProgress.invalidate(tabletId)
            return error

    @override
    getToDeleteReplicaId:
        // We don't invalidate the entry here, cuz the redundant repair progress is just started. The move should invalidate by ttl or Algo.CheckMoveCompleted()
        return movesInProgress.get(tabletId).srcReplicaId

    // Cache cleanUp may be not necessary if we cleanup cache in selectAlternativeTabletsForCluster().
    // @override
    // updateLoadStatistic:
    //     super.updateLoadStatistic()
    //     if time_interval > config:
    //         movesInProgress.cleanUp()

    @private
    // The validation check cannot be accurate, cuz the production of moves do have ordering. When some moves failed, the cluster skew is different with the skew we used in production. So we just do some basic checks.
    checkValidation:
        check if fromBe and toBe are alive
        check if fromBe.totalReplicaCount > toBe.totalReplicaCount
        check if partition.fromBe.replicaCount > 
partition.toBe.replicaCount
        // To be improved
```
We need ClusterBalanceInfo for TwoDimensionalGreedyAlgo, actually two maps. Ref kudu ClusterBalanceInfo.
```
// Balance information for a cluster,
// excluding decommissioned bes and replicas on them.
ClusterBalanceInfo{
    partitionInfoBySkew
    beByTotoalReplicaCount
}
```
The algo persudo code:
```
TwoDimensionalGreedyAlgo:
    getNextMoves:
        // copy info for applying move
        info = copy(clusterBalanceInfo)
        applyMoves(info, movesInProgress)
        for i in range(0, maxMoveNum):
            move = getNextMove(info)
            moves.add(move)
            applyMove(info, move)

    appleyMove:
        // Update the balance state in 'ClusterBalanceInfo' with the outcome of the move
        // omitted...

    getNextMove:
        maxPartitionSkew = partitionInfoBySkew.firstKey()
        maxBeSkew = maxReplicaNumOfBe - minReplicaNumOfBe
        maxSet = clusterBalanceInfo.partitionInfoBySkew.get(maxPartitionSkew)

        // Among the partitions with maximum skew, attempt to pick a partition where there is
        // a move that improves the partition skew and the cluster skew, if possible. If
        // not, attempt to pick a move that improves the partition skew. If all partitions
        // are balanced, attempt to pick a move that preserves partition balance and
        // improves cluster skew.
        findOneMove(maxSet)

    checkMovesCompleted:
        // TODO
```

### PR DESCRIPTION
**Pull request by vagetablechicken**
## Proposed changes

RebalancerType could be configured via `Config.rebalancer_type`(BeLoad, Partition). 
PartitionRebalancer is based on TwoDimensionalGreedyAlgo.
Two dims of Doris should be cluster & partition. And we only consider about the replica count, do not consider replica size.
#4845 for further details.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have create an issue on (Fix #4845), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    // The validation check cannot be accurate, cuz the production of moves do have ordering.
+    // If some moves failed, the cluster & partition skew is different to the skew when we getNextMove.
+    // So we can't do skew check.
+    // Just do some basic checks, e.g. server available.
+    private void checkMoveValidation(ReplicaMove move) throws IllegalStateException {
+        boolean fromAvailable = infoService.checkBackendAvailable(move.fromBe);
+        boolean toAvailable = infoService.checkBackendAvailable(move.toBe);
+        Preconditions.checkState(fromAvailable && toAvailable, move + "'s bes are not all available: from " + fromAvailable + ", to " + toAvailable);
+        // To be improved
+    }
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:266]: selectAlternativeTabletsForCluster method selects all tablets for maxSkew partition from highload be. If there is no check whether the move will make partition balance better, we may fall into endless loop in such case:
3 tablets on maxload be
1 tablet on minload be
we move 2 tablets from maxload be to minload be, and maxload be switchs to minload be.

Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:266]: Not all, just pick one tablet to generate one move. and apply the move to current balance info, it's in the algo implement.

Comment (gengjun-git) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:266]: OK

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 2]
```diff
@@ -1010,6 +1010,18 @@
     @ConfField(mutable = true, masterOnly = true)
     public static int max_balancing_tablets = 100;
 
+    // Rebalancer type(ignore case): BeLoad, Partition. If type parse failed, use BeLoad as default.
+    @ConfField(masterOnly = true)
+    public static String rebalancer_type = "BeLoad";
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1015]: tablet_rebalancer_type?

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 3]
```diff
@@ -1010,6 +1010,18 @@
...
+
+    // Valid only if use PartitionRebalancer.
+    @ConfField(mutable = true, masterOnly = true)
+    public static long rebalance_move_expire_after_access = 600; // 600s
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1019]: If only valid for PartitionRebalancer, the name would better add a partition prefix.

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java (hunk 2 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1015]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java (hunk 3 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1019]: ok

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+public class PartitionRebalancer extends Rebalancer {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:43]: Could copy the description in the doc to this class comment.

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+// A two-dimensional greedy rebalancing algorithm. From among moves that
+// decrease the skew of a most skewed partition, it prefers ones that reduce the
+// skew of the cluster. A cluster is considered balanced when the skew of every
+// partition is <= 1 and the skew of the cluster is <= 1.
+//
+// The skew of the cluster is defined as the difference between the maximum
+// total replica count over all bes and the minimum total replica
+// count over all bes.
+public class TwoDimensionalGreedyAlgo {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java:45]: 我觉得这个类名应该体现出 这是一个 rebalancing 相关的算法。

然后 注释 最好解释下 两维 是哪两维？

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        // Just delete the completed or problematic moves
+        if (!toDeleteKeys.isEmpty()) {
+            movesInProgress.get().invalidateAll(toDeleteKeys);
+            movesInProgressList = movesInProgressList.stream()
+                    .filter(m -> !toDeleteKeys.contains(m.tabletId)).collect(Collectors.toList());
+        }
+
+        if (movesInProgressCache.size() > Config.max_balancing_tablets) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:87]: 我们可以把move 这个概念完全去掉吗？ 改成 balancing 或者 balancing_tablets。 我们系统里面的概念应该尽可能少，保持统一。

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java (hunk 5 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java:45]: 这里可以注明是 这个类是  inspire  或者modify from kudu.

File: docs/zh-CN/administrator-guide/operation/tablet-repair-and-balance.md [hunk 7]
```diff
@@ -215,11 +215,15 @@ TabletScheduler 里等待被调度的分片会根据状态不同，赋予不同
 
 ## 副本均衡
 
-Doris 会自动进行集群内的副本均衡。均衡的主要思想，是对某些分片，先在低负载的节点上创建一个副本，然后再删除这些分片在高负载节点上的副本。同时，因为不同存储介质的存在，在同一个集群内的不同 BE 节点上，可能存在一种或两种存储介质。我们要求存储介质为 A 的分片在均衡后，尽量依然存储在存储介质 A 中。所以我们根据存储介质，对集群的 BE 节点进行划分。然后针对不同的存储介质的 BE 节点集合，进行负载均衡调度。
+Doris 会自动进行集群内的副本均衡。目前支持两种均衡策略，负载/分区。
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply docs/zh-CN/administrator-guide/operation/tablet-repair-and-balance.md:218]: 文档里最好说明 两种均衡策略是否可以随意替换。

[the remark this answers is not in the document]
Comment (vagetablechicken) [reply docs/zh-CN/administrator-guide/operation/tablet-repair-and-balance.md:218]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:87]: move不可能去掉，调度单位就是ReplicaMove。
`movesInProgressCache.size() > Config.max_balancing_tablets` 这一句如果觉得难理解，可以加点注释

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 4 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:43]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.doris.thrift.TStorageMedium;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class MovesInProgressCache {
```
[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java:33]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        public Cache<Long, Pair<PartitionRebalancer.ReplicaMove, Long>> get() {
+            return cache;
+        }
+    }
+
+    public void updateCatalog(Map<String, ClusterLoadStatistic> statisticMap, long expireAfterAccessSecond) {
+        updateCatalog(statisticMap, expireAfterAccessSecond, TimeUnit.SECONDS);
+    }
+
+    public void updateCatalog(Map<String, ClusterLoadStatistic> statisticMap, long duration, TimeUnit unit) {
```
[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java:54]: how about `updateMapping`?

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java (hunk 5 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyAlgo.java:45]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: 不是啊。 move 是个很抽象的概念。我觉得， 一个 的balance 系统是没必要出现这个概念的，你的一个move就是一个balancing的Replica。

>movesInProgressCache.size() > Config.max_balancing_tablets 这一句如果觉得难理解，可以加点注释

我不是特指这一处，而是这个PR。 

所有balancing 算法的概念应该是统一的，这样其他人可以快速理解每个balancing算法的异同。

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java]: 这个方法不是在 update cache 吗？ 你的类名是MovesInProgressCache， 而且有getCache。 所以updateCache 是不是更好理解？

[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java]: 这里是在更新map的结构，并不是要对已有的cache做什么更新。

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: PartitionRebalancer已经是个具体的rebalancer，它的任务就是从algo那里拿到需要执行的moves，然后去执行moves，管理moves，职责已经很小了，为什么还要去隐藏move的概念？

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java]: 所以这个类的设计是不是有些问题：
1 map和cache的区别是啥？为啥cache 只有get 方法， map只有update 方法？
2 我看 LoadingCache 没有定义 load方法，这里cache的更新是啥时候发生呢？

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: 所以你可以用简单的一句话清晰地定义啥是move 吗？

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java]: 这里的更新是<cluster, <meduim, cache>这个映射的更新，因为cluster目前还是可以add的，在周期性update各种信息时顺带把映射更新了，以免selectTablet时找不到cache

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: move就是为当前replica分布如何达到平衡而计算出来的迁移步骤。
algo的角度move就是<partition id, from be, to be>，partition rebalancer选择特定tablet后的move就是<tablet id, from be, to be>。

File: fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/MovesInProgressCache.java]: 好的。

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java (hunk 6 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: > algo的角度move就是<partition id, from be, to be>，partition rebalancer选择特定tablet后的move就是<tablet id, from be, to be>。

嗯，这个是好理解的。 所以我的意思我们可以命名成 balancingPartition 和 balancingTablet？ 即使一定要用move的话，也可以改成 partitionMove 和 tabletMove ?  因为 我感觉 move 这个命名太抽象了。

Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: move是算法的基本结果，还是保留吧，`ReplicaMove`这个名字改成`TabletMove`, `PartitionReplicaMove`改为`PartitionMove`吧

Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:96]: 好的。

**Review by kangkaisen — DISMISSED** [review]
+1

**Comment by killxdcj** [comment]
Will there be any balance-related modifications in the future? The new balance strategy also needs to be considered when create new partition or restore data, but I don't see the relevant changes.

**Comment by vagetablechicken** [comment]
> Will there be any balance-related modifications in the future? The new balance strategy also needs to be considered when create new partition or restore data, but I don't see the relevant changes.

It's only a rebalacer strategy,  not include the creation or some else.

**Comment by killxdcj** [comment]
> > Will there be any balance-related modifications in the future? The new balance strategy also needs to be considered when create new partition or restore data, but I don't see the relevant changes.
> 
> It's only a rebalacer strategy, not include the creation or some else.

If partition create / restore does not consider the balance strategy of the cluster, then there will be such a situation: Tablets have just been created, and then balancer finds the cluster is imbalance, and then it will do a balance on the newly created tablets.

**Comment by killxdcj** [comment]
> > > Will there be any balance-related modifications in the future? The new balance strategy also needs to be considered when create new partition or restore data, but I don't see the relevant changes.
> > 
> > 
> > It's only a rebalacer strategy, not include the creation or some else.
> 
> If partition create / restore does not consider the balance strategy of the cluster, then there will be such a situation: Tablets have just been created, and then balancer finds the cluster is imbalance, and then it will do a balance on the newly created tablets.

Another question is has the new balance strategy been verified by the production environment? We also have similar requirements and we have implemented a similar balance strategy, but now we want to switch to this community implementation.

**Comment by vagetablechicken** [comment]
> Another question is has the new balance strategy been verified by the production environment? We also have similar requirements and we have implemented a similar balance strategy, but now we want to switch to this community implementation.

Haven't yet, only in test env. And the production envs may have some differences. If you like, plz help me to test it. BTW, our team plans to use it in production env next month.

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+        // are balanced, attempt to pick a move that preserves partition balance and
+        // improves cluster skew.
+        NavigableSet<PartitionBalanceInfo> maxSet = skewMap.get(maxPartitionSkew);
+        for (PartitionBalanceInfo pbi : maxSet) {
+            Preconditions.checkArgument(!pbi.beByReplicaCount.isEmpty(), "no information on replicas of " +
+                    "partition " + pbi.partitionId + "-" + pbi.indexId);
+
+            Long minReplicaCount = pbi.beByReplicaCount.keySet().first();
+            Long maxReplicaCount = pbi.beByReplicaCount.keySet().last();
+            LOG.info("balancing partition {}-{} with replica count skew {} (min_replica_count: {}, max_replica_count: {})",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java:181]: debug is better?

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+                    pbi.partitionId, pbi.indexId, maxPartitionSkew,
+                    minReplicaCount, maxReplicaCount);
+
+            // Compute the intersection of the bes most loaded for the table
+            // with the bes most loaded overall, and likewise for least loaded.
+            // These are our ideal candidates for moving from and to, respectively.
+            IntersectionResult maxLoaded = getIntersection(ExtremumType.MAX, pbi.beByReplicaCount, beByTotalReplicaCount);
+            IntersectionResult minLoaded = getIntersection(ExtremumType.MIN, pbi.beByReplicaCount, beByTotalReplicaCount);
+            LOG.info("partition-wise: min_count: {}, max_count: {}", minLoaded.replicaCountPartition, maxLoaded.replicaCountPartition);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java:190]: debug

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+                maxLoadedBe = maxLoaded.intersection.isEmpty() ? maxLoaded.beWithExtremumCount.get(0) : maxLoaded.intersection.get(0);
+            } else {
+                minLoadedBe = minLoaded.intersection.isEmpty() ? getRandomListElement(minLoaded.beWithExtremumCount)
+                        : getRandomListElement(minLoaded.intersection);
+                maxLoadedBe = maxLoaded.intersection.isEmpty() ? getRandomListElement(maxLoaded.beWithExtremumCount)
+                        : getRandomListElement(maxLoaded.intersection);
+            }
+
+            LOG.debug("min_loaded_be: {}, max_loaded_be: {}", minLoadedBe, maxLoadedBe);
+            if (minLoadedBe.equals(maxLoadedBe)) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java:216]: If we use PICK_FIRST, in the case of fewer candidate sets, will there be a high probability of the same?

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+        if (skewMap.isEmpty() || beByTotalReplicaCount.isEmpty()) {
+            return null;
+        }
+        long maxPartitionSkew = skewMap.keySet().last();
+        long maxBeSkew = beByTotalReplicaCount.keySet().last() - beByTotalReplicaCount.keySet().first();
+
+        // 1. Every partition is balanced(maxPartitionSkew<=1) and any move will unbalance a partition, so there
+        // is no potential for the greedy algorithm to balance the cluster.
+        // 2. Every partition is balanced(maxPartitionSkew<=1) and the cluster as a whole is balanced(maxBeSkew<=1).
+        if (maxPartitionSkew == 0L || (maxPartitionSkew <= 1L && maxBeSkew <= 1L)) {
```
[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java:165]: The algo should get accurate calculation results.

File: fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+                toDeleteKeys.add(move.tabletId);
+                LOG.debug("Move {} is completed. The cur dist: {}", move,
+                        invertedIndex.getReplicasByTabletId(move.tabletId).stream().map(Replica::getBackendId).collect(Collectors.toList()));
+                counterBalanceMoveSucceeded.incrementAndGet();
+            }
+        }
+    }
+
+    // Move completed: fromBe doesn't have a replica and toBe has a replica
+    private boolean checkMoveCompleted(TabletMove move) {
```
[the remark this answers is not in the document]
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/PartitionRebalancer.java:212]: It's hard to define, but when `!bes.contains(move.fromBe) && bes.contains(move.toBe)`, the move is useless. So why not to treat it as completed.

File: fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java (hunk 12 above)
Comment (vagetablechicken) [reply fe/fe-core/src/main/java/org/apache/doris/clone/TwoDimensionalGreedyRebalanceAlgo.java:216]: Now, we only use PICK_FIRST in UTs, to get the results that can be expected.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**